net/ipv4/tcp_output.c

```diff
@@ -2632,50 +2634,31 @@ void tcp_chrono_stop(struct sock *sk, const enum tcp_chrono type)
 #ifdef CONFIG_SECURITY_TEMPESTA
 
 /**
- * The next two functions are called from places: from `tcp_write_xmit`
+ * The next funtion is called from places: from `tcp_write_xmit`
  * (a usual case) and from `tcp_write_wakeup`. In other places where
  * `tcp_transmit_skb` is called we deal with special TCP skbs or skbs
  * not from tcp send queue.
  */
-static int
-tcp_tfw_sk_prepare_xmit(struct sock *sk, struct sk_buff *skb,
-			unsigned int mss_now, unsigned int *limit,
-			unsigned int *nskbs)
-{
-	if (!sk->sk_prepare_xmit || !skb_tfw_tls_type(skb))
-		return 0;
-
-	if (unlikely(*limit <= TLS_MAX_OVERHEAD)) {
-		net_warn_ratelimited("%s: too small MSS %u"
-				     " for TLS\n",
-				     __func__, mss_now);
-		return -ENOMEM;
-	}
-
-	if (*limit > TLS_MAX_PAYLOAD_SIZE + TLS_MAX_OVERHEAD)
-		*limit = TLS_MAX_PAYLOAD_SIZE;
-	else
-		*limit -= TLS_MAX_OVERHEAD;
-
-	if (unlikely(skb_tfw_flags(skb) & SS_F_HTTP2_FRAME_PREPARED)) {
-		*nskbs = 1;
-		return 0;
-	}
-
-	return sk->sk_prepare_xmit(sk, skb, mss_now, limit, nskbs);
-}
-
 static int
 tcp_tfw_sk_write_xmit(struct sock *sk, struct sk_buff *skb,
-		      unsigned int mss_now, unsigned int limit,
-		      unsigned int nskbs)
+		      unsigned int mss_now)
 {
+	struct tcp_sock *tp = tcp_sk(sk);
+	unsigned int in_flight = tcp_packets_in_flight(tp);
+	unsigned int send_win, cong_win;
+	unsigned int limit;
```

Comment:
Because we want to make one TLS record for several skbs. For example if mtu is equal 1500, our previously calculated limit is also about 1500, but if there are several skbs in our socket write queue we can encrypt them all in only one tls record! And do not call tls_encrypt several times